default.nix

```diff
@@ -19,68 +16,83 @@ with rec
+      mkNode = name: obj: {
+        inherit name obj;
+        inherit (obj) version;
+        key =
+          if builtins.hasAttr "integrity" obj
+          then "${name}-${obj.version}-${obj.integrity}"
+          else "${name}-${obj.version}-no-integrity";
+        next =
+          if builtins.hasAttr "dependencies" obj
+          then pkgs.lib.mapAttrsToList mkNode (obj.dependencies)
+          else [];
+      };
+
+      # The list of all packages discovered in the package-lock, excluding
+      # the top-level package.
+      flattened = builtins.genericClosure {
+        startSet = [ (mkNode packageLock.name packageLock) ];
+        operator = x: x.next;
+      };
+
+      # Create an entry for the snapshot, e.g.
+      #     { some-package = { some-version = { url = ...; shaX = ...} ; }; }
+      snapshotEntry = x:
+        let
+          sha =
+            if pkgs.lib.hasPrefix "sha1-" x.obj.integrity
+            then { sha1 = pkgs.lib.removePrefix "sha1-" x.obj.integrity; } else
+              if pkgs.lib.hasPrefix "sha512-" x.obj.integrity
+              then { sha512 = pkgs.lib.removePrefix "sha512-" x.obj.integrity; }
+              else abort "Unknown sha for ${x.obj.integrity}";
+        in
           if builtins.hasAttr "resolved" x.obj
           then
-            { "${x.name}" =
-                { "${x.version}" = pkgs.fetchurl ({ url = x.obj.resolved; } // sha);
-                };
+            {
+              "${x.name}" = {
+                "${x.version}" = pkgs.fetchurl ({ url = x.obj.resolved; } // sha);
+              };
             }
           else {};
-      };
-    pkgs.lib.foldl
-    (acc: x:
-      (pkgs.lib.recursiveUpdate acc (snapshotEntry x))
-    ) {} flattened;
+
+      mergeSnapshotEntries = acc: x:
+        pkgs.lib.recursiveUpdate acc (snapshotEntry x);
+    in
+      pkgs.lib.foldl
+        mergeSnapshotEntries
+        {}
+        flattened;
 
   # Returns either the package-lock or the npm-shrinkwrap. If none is found
   # returns null.
   findPackageLock = src:
-    with rec
-      { toplevel = builtins.readDir src;
-        hasPackageLock = builtins.hasAttr "package-lock.json" toplevel;
-        hasNpmShrinkwrap = builtins.hasAttr "npm-shrinkwrap.json" toplevel;
-      };
-    if hasPackageLock then src + "/package-lock.json"
-    else if hasNpmShrinkwrap then src + "/npm-shrinkwrap.json"
-    else null;
+    let
+      toplevel = builtins.readDir src;
+      hasPackageLock = builtins.hasAttr "package-lock.json" toplevel;
+      hasNpmShrinkwrap = builtins.hasAttr "npm-shrinkwrap.json" toplevel;
+    in
+      if hasPackageLock then src + "/package-lock.json"
+      else if hasNpmShrinkwrap then src + "/npm-shrinkwrap.json"
+      else null;
+
+  # Returns the package.json as nix values. If not found, returns an empty
+  # attrset.
+  readPackageJSON = src:
+    let
+      toplevel = builtins.readDir src;
+      hasPackageJSON = builtins.hasAttr "package.json" toplevel;
```

Comment:
Why not `builtins.pathExists`?

Reply:
I just copied what you had done with `findPackageLock`.